smart-contracts/test/Lock/transferFee.js

```diff
@@ -1,159 +1,123 @@
-//     })
-//   })
-// })
+const Units = require('ethereumjs-units')
+const BigNumber = require('bignumber.js')
+
+const deployLocks = require('../helpers/deployLocks')
+const shouldFail = require('../helpers/shouldFail')
+
+const unlockContract = artifacts.require('../Unlock.sol')
+const getProxy = require('../helpers/proxy')
+
+let unlock, locks
+
+contract('Lock / transferFee', accounts => {
+  let lock
+  const keyPrice = new BigNumber(Units.convert('0.01', 'eth', 'wei'))
+  const keyOwner = accounts[1]
+
+  before(async () => {
+    unlock = await getProxy(unlockContract)
+    // TODO test using an ERC20 priced lock as well
+    locks = await deployLocks(unlock, accounts[0])
+    lock = locks['FIRST']
+    await lock.purchase(0, keyOwner, web3.utils.padLeft(0, 40), [], {
+      value: keyPrice.toFixed(),
+    })
+  })
+
+  it('has a default fee of 0%', async () => {
+    const feeNumerator = new BigNumber(await lock.transferFeeBasisPoints.call())
+    const feeDenominator = new BigNumber(await lock.BASIS_POINTS_DEN.call())
+    assert.equal(feeNumerator.div(feeDenominator).toFixed(), 0.0)
+  })
+
+  describe('once a fee of 5% is set', () => {
+    let fee, fee1, fee2, fee3, nowBefore
+    before(async () => {
+      // Change the fee to 5%
+      await lock.updateTransferFee(500)
+      nowBefore = Math.floor(Date.now() / 1000)
+    })
+
+    it('estimates the transfer fee, which is 5% of remaining duration or less', async () => {
+      fee = new BigNumber(await lock.getTransferFee.call(keyOwner, 0))
+      let expiration = new BigNumber(
+        await lock.keyExpirationTimestampFor.call(keyOwner)
+      )
+      let nowAfter = Math.floor(Date.now() / 1000)
+      assert(fee.lte(expiration.minus(nowBefore).times(0.05)))
+      assert(fee.lte(expiration.minus(nowAfter).times(0.05)))
+    })
+
+    it('calculates the fee based on the time value passed in', async () => {
+      fee1 = await lock.getTransferFee.call(keyOwner, 100)
+      fee2 = await lock.getTransferFee.call(keyOwner, 60 * 60 * 24 * 365)
+      fee3 = await lock.getTransferFee.call(keyOwner, 60 * 60 * 24 * 30)
+      assert.equal(fee1, 5)
+      assert.equal(fee2, 1576800)
+      assert.equal(fee3, 129600)
+    })
+
+    describe('when the key is transfered', () => {
+      const newOwner = accounts[2]
+      let tokenId, expirationBefore, expirationAfter, fee
+
+      before(async () => {
+        tokenId = await lock.getTokenIdFor.call(keyOwner)
+        expirationBefore = new BigNumber(
+          await lock.keyExpirationTimestampFor(keyOwner)
+        )
+        fee = await lock.getTransferFee(keyOwner, 0)
+        await lock.transferFrom(keyOwner, newOwner, tokenId, {
+          from: keyOwner,
+        })
+        expirationAfter = new BigNumber(
+          await lock.keyExpirationTimestampFor(newOwner)
+        )
+      })
+
+      it('the fee is deducted from the time transferred', async () => {
+        assert(expirationAfter.gte(expirationBefore.minus(fee)))
```

Comment:
hmm shouldn't we be asserting lte? to ensure that the fee was correctly charged?  this test would pass if the fee was missed completely, right?

Reply:
Yup, it would still pass if the fee was missed. hmm.

Reply:
we may need to be doing this:
```
assert(expirationAfter.lt(expirationBefore))
assert(expirationAfter.gte(expirationBefore.minus(fee)))
```

Reply:
I think this will work:
```
        // make sure that a fee was taken
        assert(expirationAfter.lt(expirationBefore))
        // check that fee was not more than 5%
        assert(expirationAfter.gte(expirationBefore.minus(fee)))
```

Reply:
Another way to say this: The `expirationTimestamp` after should be _at least_ the `expirationTimestamp` before - fee. If it's less, then too big a fee was taken.